### ISSUE TITLE
zetaclient stops connecting to external chain RPCs after receiving HTTP error codes. 

### ISSUE THREAD

**Issue by CharlieMc0** [zeta-chain/node#3328]
**Describe the Bug**
When zetaclientd gets an HTTP error back from the BTC RPC it stops working and never retries until zetaclientd is restarted. 

**To Reproduce**
Would need to replicate the the 409 or other HTTP errors codes from the BTC RPC. It isn't clear if this is limited to BTC or an issue for the RPC connectivity for all networks. 

**Expected Behavior**
Ideally all attempts to connect to RPCs for external networks should have a backoff timer and continue to retry even if they receive errors from the external RPC server.  

**Screenshots**

<img width="1058" alt="Screenshot 2024-12-29 at 6 07 06 PM" src="https://github.com/user-attachments/assets/c9816062-0fb2-4854-b00e-bce701063683" />

You can see from the logs one one validator was falling behind on BTC blocks
<img width="610" alt="Screenshot 2024-12-29 at 6 07 32 PM" src="https://github.com/user-attachments/assets/49f20736-2bc1-403d-8121-3a2dc415d5a8" />
<img width="582" alt="Screenshot 2024-12-29 at 6 11 04 PM" src="https://github.com/user-attachments/assets/fc23af34-36e7-4ece-af32-51e9cc6f8a9f" />

**Comment by gartnera** [comment]
Both of the places were `FetchUTXOs` is called do not exit if error. So if this is really happening then the log messages are probably unrelated.

What we really need in cases like this is a goroutine dump via `curl -o goroutine.dump http://localhost:6061/debug/pprof/goroutine?debug=2`

https://github.com/zeta-chain/node/blob/80ca9214bf531668b0b18199f0743fa81957f3ca/zetaclient/chains/bitcoin/observer/observer.go#L308-L319

https://github.com/zeta-chain/node/blob/80ca9214bf531668b0b18199f0743fa81957f3ca/zetaclient/chains/bitcoin/signer/signer.go#L206-L213

**Comment by ws4charlie** [comment]
Seems we only kept 7 days of logs in Datadog and above log prints are not tracked any more. After searching above log prints in code, the initial error was on Bitcoin RPC `GetBlockCount` (error getting block height), and then it failed `FetchUTXOs`. When `FetchUTXOs` fails, zetaclient won't be able to get balance and outbound will not work properly.

@gartnera As you said, the Bitcoin RPCs are designed to NOT take a context, so they should be able to resume/reconnect by themselves. The root cause is not looking clear with above logs.

**Comment by ws4charlie** [comment]
The whole process was:

- We have a `watchRPCStatus` goroutine that NEVER stops for Bitcoin chain.
- Every mininute, this goroutine either print `RPC latency is OK` or an error message.
- The last heart beat of `validator1-us-east-1-mainnet` was on `16:20:43` as shown in below screenshot.
- The heart beat resumed after restarting zetaclient `18:58:33`.

Here are my guesses after investigation into logs.

1. The `watchRPCStatus` go-routine got stuck on this call [blockTime, err := rpc.CheckRPCStatus(ob.btcClient, tssAddress)](https://github.com/zeta-chain/node/blob/release%2Fv24/zetaclient/chains/bitcoin/observer/rpc_status.go#L38) for 2.5 hours.
2. The `ObserveInbound` go-routine also got stuck, for similar reason, on this line [err := ob.ObserveInbound(ctx)](https://github.com/zeta-chain/node/blob/release%2Fv24/zetaclient/chains/bitcoin/observer/inbound.go#L53)

The reason is that both go-routines either **NEVER** work silently, they print message on every iteration. To be specific,
`watchRPCStatus` prints `RPC latency is OK` or error;
`ObserveInbound` prints `observeInboundBTC: block XXX has YYY txs`
We see no logs like above after last heart beat.


On which lines the above two go-routines got stuck?
A go routine stops execution only if it is blocked by a lock or IO. There is no lock used in above go-routines so the IO is what we can suspect.

In our case, the RPC status check got stuck on one of the following RPC calls in function [CheckRPCStatus](https://github.com/zeta-chain/node/blob/release%2Fv24/zetaclient/chains/bitcoin/rpc/rpc.go#L185)
`GetBlockCount, GetBlockHash, GetBlockHeader, ListUnspentMinMaxAddresses`.

The inbound go routine got stuck on one of the two lines (not sure which one):
- `currentBlock, err := ob.btcClient.GetBlockCount()` at [line 76](https://github.com/zeta-chain/node/blob/release%2Fv24/zetaclient/chains/bitcoin/observer/inbound.go#L76)  that invokes `GetBlockCount`.
- `res, err := ob.GetBlockByNumberCached(int64(blockNumber))` at [line 108](https://github.com/zeta-chain/node/blob/release%2Fv24/zetaclient/chains/bitcoin/observer/inbound.go#L108), which invokes `GetBlockHash, GetBlockHeader, GetBlockVerboseTx` internally.

Restarting the program solved the RPC errors was a interesting hint. The `409` error indicates `a request conflict with the current state of the target resource` and was resolved by a restart?


The link to [Grafana logs](https://grafana.prod-zetachain-internal.com/a/grafana-lokiexplore-app/explore/systemd_service/zetaclientd/logs?from=2024-12-29T21:55:00.000Z&to=2024-12-30T01:05:00.000Z&var-ds=loki&var-filters=systemd_service%7C%3D%7Czetaclientd&var-filters=service_name%7C%3D%7Cvalidator0-us-east-1-mainnet&var-fields=chain%7C%3D%7C%7B%22value%22:%228332%22__gfc__%22parser%22:%22structuredMetadata%22%7D,8332&var-levels=&patterns=%5B%5D&var-metadata=&var-patterns=&var-lineFilter=&urlColumns=%5B%22timestamp%22,%22body%22%5D&visualizationType=%22logs%22&displayedFields=%5B%5D&var-labelBy=$__all&var-fieldBy=$__all&refresh=)
![image](https://github.com/user-attachments/assets/c2c50f9f-e5cb-4d14-83a5-99cf5dedc66b)

**Comment by CharlieMc0** [comment]
What's a good way to avoid this in the future? We need to ensure zetaclient can handle unusual responses from RPCs without going offline. 

 Something like a RPC heartbeat fails to work after 5 minutes restart the go-routine for that network?

**Comment by ws4charlie** [comment]
This feels to restarting the zetaclient reset/released a process-level resource (don't know what was it), therefor solved the RPC hung problem. The above issue was like (for instance):
`zetaclient` go routine calls `GetBlockCount` and the call just got stuck right on the line and never returned. So the program execution was unexpected hung (the Bitcoin RPCs were designed return, without a context). A remedy to this would be external and operational in stead of self diagnosis inside zetaclient.

We would leave this issue under monitoring for a while before taking effective actions so we can collect and identify some patterns/clues of the errors.

**Comment by gartnera** [comment]
> What's a good way to avoid this in the future? We need to ensure zetaclient can handle unusual responses from RPCs without going offline.

Completely reimplement the bitcoin rpc such that it takes contexts and properly returns on timeout/context cancelation.

Medium to high effort

**Comment by swift1337** [comment]
> Completely reimplement the Bitcoin RPC such that it takes contexts and properly returns on timeout/context cancellation.

I agree, this is the only reliable way of solving this.

We need to write a wrapper around `github.com/btcsuite/btcd/btcjson` that uses the same DTOs but has proper HTTP & ctx support. Only used methods should be implemented, no need to code all supported RPC methods.

We can also leverage this moment to bring proper retries, logs, and Prometheus metrics.

**Comment by ws4charlie** [comment]
@gartnera I agree with the short/medium term to setup watchdog to automatically restart. Let's monitor how frequently the issue would happen moving forward and find some patterns. If necessary, re-implementing (or fork and add context) Bitcoin RPC is always not a bad idea if RPC calls hang a lot.
